File: app/src/main/java/com/orozco/netreport/flux/store/DataCollectionStore.java
package com.orozco.netreport.flux.store;

import com.orozco.netreport.flux.Action;
import com.orozco.netreport.flux.AppError;
import com.orozco.netreport.flux.Store;
import com.orozco.netreport.model.Data;

import rx.Observable;

import static com.orozco.netreport.flux.action.DataCollectionActionCreator.DataCollectionAction.ACTION_COLLECT_DATA_F;
import static com.orozco.netreport.flux.action.DataCollectionActionCreator.DataCollectionAction.ACTION_COLLECT_DATA_S;
import static com.orozco.netreport.flux.action.DataCollectionActionCreator.DataCollectionAction.ACTION_SEND_DATA_F;
import static com.orozco.netreport.flux.action.DataCollectionActionCreator.DataCollectionAction.ACTION_SEND_DATA_S;

/**
 * @author A-Ar Andrew Concepcion
 */
public class DataCollectionStore extends Store<DataCollectionStore> {
    private Data data = null;
    private String action = null;
    private AppError error = null;

    public Observable<DataCollectionStore> observableWithFilter(String filter) {
        return observable().filter(store -> filter.equals(store.action));
    }

    private void updateState() {
        error = null;
    }

    private void updateData(Action action) {
        data = (Data) action.data();
    }

    private void updateError(Action action) {
        if (action.data() != null && action.data() instanceof AppError) {
            error = (AppError) action.data();
        }
    }

    private void updateAction(Action action) {
        this.action = action.type();
    }

    @Override
    protected void onReceiveAction(Action action) {
        switch (action.type()) {
            case ACTION_COLLECT_DATA_S:
            case ACTION_COLLECT_DATA_F:
            case ACTION_SEND_DATA_S:
            case ACTION_SEND_DATA_F:
                updateState();
                updateData(action);
                updateError(action);
                updateAction(action);
                notifyStoreChanged(this);
                break;
        }
    }

    public Data getData() {
        return data;
    }

    public String getAction() {
        return action;
    }

    public AppError getError() {
        return error;
    }
}

File: app/src/main/java/com/orozco/netreport/flux/store/DataCollectionStore.kt
package com.orozco.netreport.flux.store

import com.orozco.netreport.flux.Action
import com.orozco.netreport.flux.AppError
import com.orozco.netreport.flux.Store
import com.orozco.netreport.flux.action.DataCollectionActionCreator
import com.orozco.netreport.model.Data
import rx.Observable

/**
 * @author A-Ar Andrew Concepcion
 */
class DataCollectionStore : Store<DataCollectionStore>() {
    var data: Data? = null
        private set
    var action: String? = null
        private set
    var error: AppError? = null
        private set

    fun observableWithFilter(filter: String): Observable<DataCollectionStore> {
        return observable().filter { store -> filter == store.action }
    }

    private fun updateState() {
        error = null
    }

    private fun updateData(action: Action) {
        if (action.data() is Data) {
            data = action.data() as Data?
        }
    }

    private fun updateError(action: Action) {
        if (action.data() != null && action.data() is AppError) {
            error = action.data() as AppError?
        }
    }

    private fun updateAction(action: Action) {
        this.action = action.type()
    }

    override fun onReceiveAction(action: Action) {
        when (action.type()) {
            DataCollectionActionCreator.ACTION_COLLECT_DATA_S,
            DataCollectionActionCreator.ACTION_COLLECT_DATA_F,
            DataCollectionActionCreator.ACTION_SEND_DATA_S,
            DataCollectionActionCreator.ACTION_SEND_DATA_F -> {
                updateState()
                updateData(action)
                updateError(action)
                updateAction(action)
                notifyStoreChanged(this)
            }
        }
    }
}
